### ISSUE TITLE
Is using torch.no_grad() in d_unrolled_loop ignoring the unrolling?

### ISSUE THREAD

**Issue by Avsecz** [andrewliao11/unrolled-gans#1]
Hi,

thanks for making the implementation open source. In the `d_unrolled_loop`, why are the gradients stopped for the generator (using `torch.no_grad():`)?
````python
def d_unrolled_loop(d_gen_input=None):
 ....
    with torch.no_grad():
        d_fake_data = G(d_gen_input)
    d_fake_decision = D(d_fake_data)
    target = torch.zeros_like(d_fake_decision)
    d_fake_error = criterion(d_fake_decision, target)  # zeros = fake
    
    d_loss = d_real_error + d_fake_error
    d_loss.backward(create_graph=True)
    d_optimizer.step()     # Only optimizes D's parameters; 
    return d_real_error.cpu().item(), d_fake_error.cpu().item()
````

I namely think that the gradient update of G in `g_loop` should take into account the un-rolling of the discriminator training. E.g. we should compute the gradient w.r.t. Generator weights during the un-rolling procedure.:

```python
def g_loop():
    # 2. Train G on D's response (but DO NOT train D on these labels)
    g_optimizer.zero_grad()

    gen_input = torch.from_numpy(noise_sampler(minibatch_size, g_inp))
   if unrolled_steps > 0:
        backup = copy.deepcopy(D)
        for i in range(unrolled_steps):
            d_unrolled_loop(d_gen_input=gen_input)
    
    g_fake_data = G(gen_input)
    dg_fake_decision = D(g_fake_data)
    target = torch.ones_like(dg_fake_decision)
    if cuda:
        target = target.cuda()
    g_error = criterion(dg_fake_decision, target)  # we want to fool, so pretend it's all genuine
    g_error.backward()
    g_optimizer.step()  # Only optimizes G's parameters
    
    if unrolled_steps > 0:
        D.load(backup)    
        del backup
```
